lib/datadog/tracing/contrib/opensearch/patcher.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+require_relative '../../metadata/ext'
+require_relative '../analytics'
+require_relative 'ext'
+require_relative '../ext'
+require_relative '../integration'
+require_relative '../patcher'
+
+module Datadog
+  module Tracing
+    module Contrib
+      module OpenSearch
+        # Patcher enables patching of 'opensearch' module.
+        module Patcher
+          include Contrib::Patcher
+
+          module_function
+
+          def patch
+            require 'uri'
+            require 'json'
+            require_relative 'quantize'
+
+            ::OpenSearch::Transport::Client.prepend(Client)
+          end
+
+          # Patches OpenSearch::Transport::Client module
+          module Client
+            # rubocop:disable Metrics/MethodLength
+            def perform_request(*args)
+              service = datadog_configuration[:service_name]
+              method = args[0]
+              path = args[1]
+              params = args[2]
+              body = args[3]
+              response = nil
+              Tracing.trace('opensearch.query', service: service) do |span|
+                begin
+                  # set generic tags
+                  span.set_tag(Tracing::Metadata::Ext::TAG_COMPONENT, Ext::TAG_COMPONENT)
+                  span.set_tag(Tracing::Metadata::Ext::TAG_KIND, Tracing::Metadata::Ext::SpanKind::TAG_CLIENT)
+                  span.set_tag(Contrib::Ext::DB::TAG_SYSTEM, Ext::TAG_SYSTEM)
+
+                  # set argument tags
+                  span.set_tag(Datadog::Tracing::Contrib::OpenSearch::Ext::TAG_METHOD, method)
```

Comment:
minor comment:
```suggestion
                  span.set_tag(OpenSearch::Ext::TAG_METHOD, method)
```

The way Ruby searches for constants (constants are anything starting with an uppercase letter) is by starting from the let-most part (in this case `Datadog`, for `Datadog::Tracing::Contrib::OpenSearch::Ext::TAG_METHOD`), and starts walking up the module/class declaration where this is declared.

In this case, it's:
```
module Datadog
  module Tracing
    module Contrib
      module OpenSearch
        module Patcher
          module Client
```

Ruby will look for `Datadog` (from `Datadog::Tracing::Contrib::OpenSearch::Ext::TAG_METHOD`), in `Client::Datadog` but not find it, then `Patcher::Datadog` but not find it, then `OpenSearch::Datadog` but not find it and so on. It will eventually find `Datadog` as a root module.
When `Datadog` is found, it will then navigate from left to right finding each nested constant from `Datadog::Tracing::Contrib::OpenSearch::Ext::TAG_METHOD`, until it reaches `TAG_METHOD`.

At the end of day, what we really want here is `TAG_METHOD`, we don't care about all the rest of the hierarchy.
Because we only need `TAG_METHOD` here, we can find it easier in the class hierarchy by shortening our constant lookup to: `OpenSearch::Ext::TAG_METHOD`.

This means that Ruby will look for `OpenSearch` up the current module/class declaration, and it will find it here:
```
module Datadog
  module Tracing
    module Contrib
      module OpenSearch # <--
        module Patcher
          module Client
```

Which will translate to `Datadog::Tracing::Contrib::OpenSearch`
Then Ruby will try to follow the constant lookup from left to right, looking up `Ext` inside `Datadog::Tracing::Contrib::OpenSearch`, which is correct, then finding the correct `TAG_METHOD`.

The rule here is to always declare a common ancestor, in this case `OpenSearch` is the common ancestor.

As a side note, one might be tempted to simply declare `TAG_METHOD`, but the issue is that `TAG_METHOD` is not directly nested to any of the scoped modules:
```
module Datadog
  module Tracing
    module Contrib
      module OpenSearch
        module Patcher
          module Client
```
Thus lookup will fail.

`Ext::TAG_METHOD` would actually work here, but leaves it ambiguous where in the hierarchy chain `Ext` was found.

```
module Datadog # was `Ext` found here?
  module Tracing # or here?
    module Contrib # maybe here?
      module OpenSearch # or here?
        module Patcher # here maybe?
          module Client # probably not here?
```

Adding the extra `OpenSearch` to it allows us to easily find `OpenSearch` in the hierarchy and know that's our anchor point.

As a second side note, if we need to refer to a global module, but we have a module with the name name in our current hierarchy, we prefix it with `::`. In our case, simply writing `OpenSearch` will retrieve us `Datadog::Tracing::Contrib::OpenSearch`, not the `OpenSearch` client from `opensearch-ruby`. To get that client, we need to do `::OpenSearch`. This is only the case because are inside `Datadog::Tracing::Contrib::OpenSearch` in this file.